### ISSUE TITLE
Feature request: check property names in match case 

### ISSUE THREAD

**Issue by arjenzorgdoc** [microsoft/pyright#7179]
**Is your feature request related to a problem? Please describe.**

The problem is that I cannot be bother to type my attribute names correctly:

```python
# pyright: strict
from dataclasses import dataclass

@dataclass
class Point:
    x: int
    y: int


p = Point(0, 0)

match p:
    case Point(ex=0):   # <- bug here, not detected by pyright
        print("p is on the y axis")
    case Point(ex=pos):  # Type of "pos" is unknown -> error correctly detected by pyright
        print(f"p is on {pos=}")
    case _:
        print("?")
```

**Describe the solution you’d like**

I'd like pyright to warn me about wrong attribute names in match case statements.

```
  /tmp/point.py:21:9 - error: Cannot access member "ex" for type "Point"
    Member "ex" is unknown (reportAttributeAccessIssue)
```

**Comment by erictraut** [comment]
When I initially implemented support for `match` in pyright, I didn't implement a check for an unexpected property name because of the way that `__match_args__` works. It can be specified in a way that a type checker cannot determine its value statically. In such cases, a static type checker cannot determine whether a referenced property name is valid or not. I wanted to avoid false positive errors that could result from this situation. However, when `__match_args__` is auto-generated, as it is with dataclasses, there should be no ambiguity or source of false positives. I think this is a reasonable feature request in those cases at least.

**Comment by erictraut** [comment]
Actually, there's another issue here. A class pattern match performs an `isinstance` check, which means that subclasses of the class specified in the class pattern will also be matched. Those subclasses could potentially add new attributes that are not defined in the parent class. That means a type checker cannot issue an error under this condition.

Consider the following example:
```python
@dataclass
class A:
    pass

@dataclass
class B(A):
    b: int

def func(subj: A):
    match subj:
        case A(b=0):
            print("here")

func(B(0))
```

This means the feature you've requested won't work because it will lead to false positive errors.

**Comment by Azureblade3808** [comment]
I think
```python
def func(subj: A):
    match subj:
        case A(b=0):
            print("here")
```
is roughly equivalent to
```python
def func(subj: A):
    if isinstance(subj, A) and subj.b == 0:
        print("here")
```
, and the latter gets an error because `b` is not a known member of `A`. Why should they be treated differently?

**Comment by erictraut** [comment]
They aren't equivalent because the first case uses pattern matching, and if the pattern match fails, the runtime simply moves on to the next case statement. No exception is raised if `b` doesn't exist.

This is a perfectly legitimate and reasonable use of class patterns:

```python
class A:
    pass

class B(A):
    def __init__(self, value: int):
        self.b = value

def func(subj: A):
    match subj:
        case A(b=0):
            print("A with b=0")
        case A(b=x):
            print(f"A with b={x}")
        case A():
            print("A without b")

func(B(0)) # prints "A with b=0"
func(B(1)) # prints "A with b=1"
func(A()) # prints "A without b"
```

**Comment by Azureblade3808** [comment]
Got it. Thank you for explaining.

**Comment by arjenzorgdoc** [comment]
This also is perfectly fine python, but it's something we don't want:

```python
def set_b(a: A):
    a.b = 12
```

I would prefer this:

```python
class A:
    pass


class B(A):
    def __init__(self, value: int):
        self.b = value


class C(A):
    def __init__(self, value: int):
        self.b = value


def set_b1(a: B | C):
    a.b = 1


def set_b2(a: A):
    setattr(a, "b", 2)


def set_b3(a: A):
    a.b = 3  # type: ignore


def func(subj: A):
    match subj:
        case B(b=1) | C(b=1):
            print("case 1")
        case A() if getattr(subj, "b", None) == 2:
            print("case 2")
        case A(b=3):   # type: ignore
            print("case 3")
```

`set_b1` and "case 1" are tedious. You have to specify all the types that have a `b`. 

`set_b2` and "case 2": use `getattr`/`setattr` to silence the type checker. This is silly. Just use:

`set_b3` and "case 3": because it never fails at runtime, just use `# type ignore`
